kimchi/src/circuits/gates/poseidon.rs

```diff
@@ -181,3 +184,52 @@ impl<F: FftField> CircuitGate<F> {
         })
     }
 }
+
+/// `generate_witness(row, params, witness_cols, input)` uses a sponge initialized with
+/// `params` to generate a witness for starting at row `row` in `witness_cols`,
+/// and with input `input`.
+pub fn generate_witness<F: Field>(
```

Comment:
I like this generation of the witness being part of the gate itself, rather than part of the tests only. Are you planning to extend this refactoring to other gates as well, even for the ones whose witness generation isn't as complex?

Reply:
imo it'd be really great if that was the case, but I haven't thought too much about it. For it to be really useful it'd have to be something that can be used by the OCaml code. [Right now the OCaml code assembles the execution trace/witness](https://github.com/MinaProtocol/mina/blob/develop/src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml#L941), but it'd be nice if it could just call a function with the right parameters. But for that to work we'd have to create an execution trace/witness_cols struct

Reply:
imo it'd be really great if that was the case, but I haven't thought too much about it. For it to be really useful it'd have to be something that can be used by the OCaml code. [Right now the OCaml code assembles the execution trace/witness](https://github.com/MinaProtocol/mina/blob/develop/src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml#L941), but it'd be nice if it could just call a function with the right parameters. But for that to work we'd have to create an execution trace/witness_cols struct

Reply:
Oh I see... Actually I was not sure where the witness was being generated, so it is in OCaml I see... In any case, it seems like this is being a thing in Cairo and ForeignMul, but these are examples of quite complex witness generations.

Reply:
yeah which is not great IMO, because if someone wants to use this library they'll have no idea how to do that :) I'm actually writing something that could be a start atm